tests/query/joins/test_equivalences.py

```diff
@@ -0,0 +1,346 @@
+                    rhs_entity=EntityKey.GROUPASSIGNEE,
+                    keys=[
+                        JoinCondition(
+                            JoinConditionExpression("", "user_id"),
+                            JoinConditionExpression("", "user_id"),
+                        )
+                    ],
+                    join_class=JoinClass.ONE_2_N,
+                    equivalences=[],
+                ),
+            },
+            writable_storage=None,
+        )
+
+
+class GroupAssignee(FakeEntity):
+    def __init__(self) -> None:
+        super().__init__(
+            storages=[],
+            query_pipeline_builder=Mock(),
+            abstract_column_set=GROUPS_ASSIGNEE,
+            join_relationships={
+                "events": JoinRelationship(
+                    rhs_entity=EntityKey.EVENTS,
+                    keys=[
+                        JoinCondition(
+                            JoinConditionExpression("", "group_id"),
+                            JoinConditionExpression("", "group_id"),
+                        )
+                    ],
+                    join_class=JoinClass.ONE_2_N,
+                    equivalences=[ColumnEquivalence("project_id", "project_id")],
+                ),
+                "user_assigned": JoinRelationship(
+                    rhs_entity=EntityKey.EVENTS,
+                    keys=[
+                        JoinCondition(
+                            JoinConditionExpression("", "user_id"),
+                            JoinConditionExpression("", "user_id"),
+                        )
+                    ],
+                    join_class=JoinClass.ONE_2_N,
+                    equivalences=[ColumnEquivalence("project_id", "project_id")],
+                ),
+            },
+            writable_storage=None,
+        )
+
+
+FULL_PROJECT_ID_COMPONENT = {
+    QualifiedCol(EntityKey.GROUPEDMESSAGES, "project_id"),
+    QualifiedCol(EntityKey.GROUPASSIGNEE, "project_id"),
+    QualifiedCol(EntityKey.EVENTS, "project_id"),
+}
+
+TEST_CASES = [
+    pytest.param(
+        JoinClause(
+            IndividualNode("ev", EntitySource(EntityKey.EVENTS, EVENTS_SCHEMA, None)),
+            IndividualNode(
+                "gr", EntitySource(EntityKey.GROUPEDMESSAGES, GROUPS_SCHEMA, None)
+            ),
+            [
+                JoinCondition(
+                    JoinConditionExpression("ev", "group_id"),
+                    JoinConditionExpression("gr", "id"),
+                )
+            ],
+            JoinType.INNER,
+            None,
+        ),
+        {
+            QualifiedCol(EntityKey.EVENTS, "group_id"): {
+                QualifiedCol(EntityKey.GROUPEDMESSAGES, "id"),
+                QualifiedCol(EntityKey.EVENTS, "group_id"),
+            },
+            QualifiedCol(EntityKey.GROUPEDMESSAGES, "id"): {
+                QualifiedCol(EntityKey.EVENTS, "group_id"),
+                QualifiedCol(EntityKey.GROUPEDMESSAGES, "id"),
```

Comment:
I feel like the column doesn't need to point to itself. Is there a reason you are tracking that?

Reply:
I left it in the connected component for each node because a connected component is the set of all nodes that are connected, not all the nodes reachable from node `x` and I had a hard time finding a concise naming for the "nodes reachable from a node".
That's the only reason, no strong opinion. I can remove if this is confusing

Reply:
It seems a little counter-intuitive. As in, if no other entities are reachable from an entity, len(equivalencies) == 1 instead of 0.

Reply:
ok, will change.